utils/cp_parallel.py

```diff
@@ -0,0 +1,133 @@
+        plate_name = result.args[6].name
+        output_string = result.stderr.decode("utf-8")
+
+        # set log file name as plate name from command
+        log_file_path = pathlib.Path(f"{log_dir}/{plate_name}_{run_name}_run.log")
+        # print output to a log file for each plate to view after the run
+        with open(log_file_path, "w") as log_file:
+            log_file.write(plate_name + "\n")
+            log_file.write(output_string + "\n")
+
+
+def run_cellprofiler_parallel(
+    plate_info_dictionary: dict,
+    run_name: str,
+) -> None:
+    """
+    This function utilizes multi-processing to run CellProfiler pipelines in parallel.
+
+    Args:
+        plate_info_dictionary (dict): dictionary with all paths for CellProfiler to run a pipeline
+        run_name (str): a given name for the type of CellProfiler run being done on the plates (example: whole image features)
+
+    Raises:
+        FileNotFoundError: if paths to pipeline and images do not exist
+    """
+    # create a list of commands for each plate with their respective log file
+    commands = []
+
+    # make logs directory
+    log_dir = pathlib.Path("./logs")
+    os.makedirs(log_dir, exist_ok=True)
+
+    # iterate through each plate in the dictionary
+    for _, info in plate_info_dictionary.items():
+        # set paths for CellProfiler
+        path_to_pipeline = info["path_to_pipeline"]
+        path_to_images = info["path_to_images"]
+        path_to_output = info["path_to_output"]
+
+        # check to make sure paths to pipeline and directory of images are correct before running the pipeline
+        if not pathlib.Path(path_to_pipeline).resolve(strict=True):
+            raise FileNotFoundError(
+                f"The file '{pathlib.Path(path_to_pipeline).name}' does not exist"
+            )
+        if not pathlib.Path(path_to_images).is_dir():
+            raise FileNotFoundError(
+                f"Directory '{pathlib.Path(path_to_images).name}' does not exist or is not a directory"
+            )
+        # make output directory if it is not already created
+        pathlib.Path(path_to_output).mkdir(exist_ok=True)
+
+        # creates a command for each plate in the list
+        command = [
+            "cellprofiler",
+            "-c",
+            "-r",
+            "-p",
+            path_to_pipeline,
+            "-o",
+            path_to_output,
+            "-i",
+            path_to_images,
+        ]
+        # creates a list of commands
+        commands.append(command)
+
+    # set the number of CPUs/workers as the number of commands
+    num_processes = len(commands)
+
+    # make sure that the number of workers does not exceed the maximum number of workers for the machine
+    if num_processes > multiprocessing.cpu_count():
+        raise MaxWorkerError(
+            "Exception occurred: The number of commands exceeds the number of CPUs/workers. Please reduce the number of commands."
+        )
```

Comment:
CellProfiler headless defaults to only using one core, so I don't know if it is advantageous to have all cores running. I am not sure how to check what CellProfiler is actually doing when running so maybe this would make sense.

Do you have any advice on how to check to see if running specifying all cores versus one core per process is better?

Reply:
Hey @jenna-tomkinson, this question is more about enhancing the adaptability of your program, especially when users provide incorrect parameters. Initially, I'm thinking: "What if a user doesn't know how many cores their computer has? If they exceed the available cores, should it raise an error, or should it trigger a warning indicating that it defaulted to the maximum number of cores (or a safer default like `max_cores / 2`)?"

> Do you have any advice on how to determine if running with all cores versus one core per process is better? 

My initial benchmark would involve measuring the execution time for both scenarios: one running with all cores and another running with a single core. Another valuable approach is to identify the point where diminishing returns begin. For instance, at what number of cores does the performance plateau, showing a flat trend in runtime as more cores are added? This insight can guide users in choosing an optimal number of cores for their specific use case.